text/0323-array-functions.md

```diff
@@ -0,0 +1,130 @@
+- Start Date: 2018-03-25
+- RFC PR: https://github.com/emberjs/rfcs/pull/323
+- Ember Issue: (leave this empty)
+
+# Array functions
+
+## Summary
+
+This RFC proposes a way to work with arrays in ember without requiring that we extend the global `Array` prototype or manually wrap arrays with `Ember.A(array)`.
+
+Specifically, for each method on the `MutableArray` mixin we will provide a similarly named function that takes an array as its first argument followed by the arguments to the method.
+
+For example, instead of calling `array.pushObject(item)` you can call `pushObject(array, item)`. For a full list see the expandible list in [Detailed design](#detailed-design).
+
+## Motivation
+
+Currently it is not a great experience to write an app with `Array` prototype extensions disabled. You end up needing to wrap arrays everywhere with `Ember.A(array)` which loops through each method on the `NativeArray` mixin and assigns it as an enumerable own property on the array.
+
+The are several problems here:
+
+- It's annoying to defensively write `Ember.A()` around every array
+- It's wasteful to looping through all the array methods and assign them all. Especially if you're only going to use one of them
+- Iterating over an array with `for..in` or `Object.keys` will not behave as generally expected because the methods are assigned enumerably (for performance reasons)
+
+All of these problems go away when using array functions instead of array methods.
+
+It's also worth noting that when you using fastboot you are required to have `Array` prototype extensions disabled currently and thus forced to deal with these issues.
```

Comment:
it might be good to mention that/if these new functions are compatible with the proposed [pipeline operator](http://tc39.github.io/proposal-pipeline-operator/)

Reply:
If you grep through https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/mixins/array.js for `return this` you'll see that several of the existing methods do so I guess the answer is "kind of".

The new functions would have the same inputs/outputs (aside from passing the array in the first argument).